### PR TITLE
Adding diagnostics, saving data and improving mem

### PR DESCRIPTION
**Pull request by Gulfa**

### CODE REVIEW

File: R/aaGlobalVariables.R [hunk 1]
```diff
@@ -47,20 +47,43 @@ CONFIG$DB_CONFIG <- list(
...
+        alertExternal = FALSE,
+        websiteInternal = TRUE,
+        contactType = list("Legekontakt"),
+        syndromeOrConsult = "syndrome",
+        denominator = "consult_with_influenza",
+        age = jsonlite::toJSON(list("Totalt"=c("Totalt"))),
+        weeklyDenominatorFunction = "sum",
+        multiplicative_factor = 100,
+        namesLong = "Influensa",
+        include_as_sydromes = TRUE,
```
Comment (raubreywhite) [inline R/aaGlobalVariables.R]: `include_as_sydromes` should be `include_as_syndromes`

File: R/aaGlobalVariables.R [hunk 2]
```diff
@@ -47,20 +47,43 @@ CONFIG$DB_CONFIG <- list(
...
+        alertExternal = FALSE,
+        websiteInternal = TRUE,
+        contactType = list("Legekontakt"),
+        syndromeOrConsult = "syndrome",
+        denominator = "pop",
+        age = jsonlite::toJSON(list("0-4"=c("0-4"),"5-14"=c("5-14"),
+                                    "15-64"=c("20-29", "30-64"), "65+" = c("65+"))),
+        weeklyDenominatorFunction = "mean",
+        multiplicative_factor = 100000,
+        include_as_sydromes = FALSE,
```
Comment (raubreywhite) [inline R/aaGlobalVariables.R]: `include_as_sydromes` should be `include_as_syndromes`

File: R/quasipoisson_diagnostics.R [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    db_table = glue::glue("spuls_qp_diagnostics"),
+    db_field_types = qp_diagnostics_field_types,
+    db_load_folder = "/xtmp/",
+    keys = qp_diagnostics_keys 
+  ))
+}
+
+#'
+#' @export
+extract_daignostics <- function(fit){
```
Comment (raubreywhite) [inline R/quasipoisson_diagnostics.R]: `extract_diagnostics`

File: R/r6_mem.R [hunk 4]
```diff
@@ -42,70 +49,104 @@ MeM <- R6::R6Class(
...
 
-  out <- rbind(national, counties)
-  out[, tag := conf$tag]
-  out[, rate := n / consult_with_influenza * 100]
-  out[, n := NULL]
-  out[, consult_with_influenza := NULL]
-  out[fhidata::days, on = "yrwk", date := mon]
+    data <- data.table()
+    for (part_age in data_age){
+      f = fhi::DashboardFolder(
```
Comment (raubreywhite) [inline R/r6_mem.R]: use `fd::path` instead of `fhi::DashboardFolder`

Comment (raubreywhite) [reply R/r6_mem.R]: (Any `fhi::Dashboard*` function is outdated and should be replaced with an `fd::` function)

File: R/r6_mem.R [hunk 5]
```diff
@@ -224,10 +266,10 @@ create_plots <- function(conf, mem_schema = NULL) {
   }
 }
 
-prepare_data_frame <- function(data) {
+prepare_data_frame <- function(data, mult_factor=100) {
   useful_data <- data[week %in% c(1:20, 40:52)]
   AddXToWeekly(useful_data)
```
Comment (raubreywhite) [inline R/r6_mem.R]: I think we can use `fhi::x()` instead of `AddXToWeekly`?

File: R/r6_quasipoisson.R [hunk 6]
```diff
@@ -163,41 +170,48 @@ quasi_run_age <- function(
...
   })
-
   data <- readRDS(file = file.path(base_folder, glue::glue("{latest_id}_{conf$tag}_{age}_cleaned.RDS")))
   load_stack_schema(conf = conf, data = data, schema = stack_x)
 
-  run_stack <- stack_x$get_data_dt()[exists_in_db == FALSE]
+  run_stack <- stack_x$get_data_dt()[exists_in_db == FALSE][]
   run_stack <- split(run_stack, seq(nrow(run_stack)))
 
+  diagnostics <- new_diagnostics_df()
```
Comment (raubreywhite) [inline R/r6_quasipoisson.R]: Lines 185, 202, are problematic for a few reasons.

1. It is generally a bit dangerous to use `<<-` as it is referencing things outside of the variable's scope, and not recommended.
2. R is very inefficient with memory allocation. So `diagnostics <- rbind(diagnostics, new_df)` actually reallocates new memory and copies *everything* to a new data.frame each time it runs. So this is really slow and memory inefficient. The recommended way of doing this kind of stuff is:
```
results <- vector("list", length=10)
for(i in 1:10){
results[[i]] <- new_df
}
results <- rbindlist(results)
```
Because `list`s can be edited by reference.
3. `*apply` functions should not be treated like a loop. I would recommend doing something like this:

```
a <- 1:10

results <- lapply(a, function(x){
  return(list(
    a=1,
    b=2
  ))
})

results
a <- lapply(results, function(x) x$a)
b <- lapply(results, function(x) x$b)

rbindlist(a)
rbindlist(b)
```
Basically, your `pblapply` function returns a list containing two elements. Then you pull out each of the elements to create your two proper lists, and then `rbindlist` them together.

File: R/std_QuasipoissonTestPredictData.r [hunk 7]
```diff
@@ -253,5 +255,6 @@ QuasipoissonTrainPredictData <- function(
   }
   datasetPredict[, uuid := uuid]
 
-  return(datasetPredict[, VARS$REQ_RESULTS_BASIC, with = F])
+  return(list(dataset_predict=datasetPredict[, VARS$REQ_RESULTS_BASIC, with = F],
```
Comment (raubreywhite) [inline R/std_QuasipoissonTestPredictData.r]: In this case, it seems a bit messy to change a "predict" function to returning two objects. 99% of the time, we only want `dataset_predict`. So it makes it a bit annoying to always have to think about `diagnostics`.

We can avoid this by using attributes:
```
retval <- datasetPredict[, VARS$REQ_RESULTS_BASIC, with = F]
attr(retval , "diagnostics") <- regression_diagnostics
return(retval)
```

Then you can access `diagnostics` via `attr(value, "diagnostics")`.

I think this is a cleaner method of getting this information out of the function.

Comment (Gulfa) [reply R/std_QuasipoissonTestPredictData.r]: Agreed

File: R/r6_quasipoisson.R (hunk 6 above)
Comment (Gulfa) [reply R/r6_quasipoisson.R]: Happy to go with version 3.

File: R/r6_mem.R (hunk 5 above)
Comment (Gulfa) [reply R/r6_mem.R]: That is very possible, you were the one who added AddXToWeekly in the first place, so will just try it with fhi::x()

File: R/std_QuasipoissonTestPredictData.r (hunk 7 above)
Comment (raubreywhite) [reply R/std_QuasipoissonTestPredictData.r]: You might also have to change some of the unit tests (strip out the attributes when checking equality of the results to the expected) if you're changing this function.

Comment (Gulfa) [reply R/std_QuasipoissonTestPredictData.r]: Seems like the tests in test_QuassipoissonAlgorithm.R passes, can not see any failing unittests when running devtools::test(). Should I check anything else?

**Comment by raubreywhite** [comment]
Nope. if it passes, then it is all good!

File: R/r6_mem.R [hunk 8]
```diff
@@ -268,7 +271,7 @@ create_plots <- function(conf, mem_schema = NULL) {
 
 prepare_data_frame <- function(data, mult_factor=100) {
   useful_data <- data[week %in% c(1:20, 40:52)]
-  AddXToWeekly(useful_data)
+  useful_data[, x:=fhi::x(useful_data[, week])]
```
Comment (raubreywhite) [inline R/r6_mem.R]: Doesn't this work?

```
useful_data[, x:=fhi::x(week)]
```

File: R/r6_mem.R [hunk 9]
```diff
@@ -224,10 +269,10 @@ create_plots <- function(conf, mem_schema = NULL) {
   }
 }
 
-prepare_data_frame <- function(data) {
+prepare_data_frame <- function(data, mult_factor=100) {
   useful_data <- data[week %in% c(1:20, 40:52)]
-  AddXToWeekly(useful_data)
-  useful_data[, rate := n / consult_with_influenza * 100]
+  useful_data[, x:=fhi::x(useful_data[, week])]
```
Comment (raubreywhite) [inline R/r6_mem.R]: Doesn't this work?

```
useful_data[, x:=fhi::x(week)]
```

Comment (Gulfa) [reply R/r6_mem.R]: That works, pushed this change now
